tests/benchmarks.cpp

```diff
@@ -0,0 +1,147 @@
+#include "benchmark/benchmark.h"
+#include "gmock/gmock.h"
+
+#include "mocks.h"
+
+#include "ast/passes/clone.h"
+#include "ast/passes/codegen_llvm.h"
+#include "ast/passes/field_analyser.h"
+#include "ast/passes/resource_analyser.h"
+#include "ast/passes/semantic_analyser.h"
+
+#include "bpffeature.h"
+#include "bpftrace.h"
+#include "clang_parser.h"
+#include "driver.h"
+
+namespace bpftrace::test {
+
+// We break the benchmarks into several phases, in order to track issues
+// separately. Note that currently each meaningful pass has its own phase, but
+// these could be logically merged as "optimizations" or some other passes with
+// some minor code restructuring.
+enum class Phase {
+  Parse,
+  FieldAnalyser,
+  ClangParser,
+  SemanticAnalyser,
+  ResourceAnalyser,
+  Codegen,
+};
+
+template <Phase P, Phase Is, typename Func>
+void measure(benchmark::State &state, Driver &driver, Func fn)
+{
+  // Run once only if we are not in the benchmark.
+  if constexpr (P != Is) {
+    fn();
+    return;
+  }
+
+  auto orig = std::move(driver.ctx);
+  for (auto _ : state) {
+    {
+      // Clone the AST in its current exact form. This should be relatively
+      // inexpensive (at least on the order of the pass itself), but we need to
+      // exclude this time.
```

Comment:
Implicit is that the pass may be making modifications. And that we want to capture the cost of those modifications, right? Might be worth documenting this in the comment